### PR TITLE
Get dbadatabase feature lastused

### LINKED ISSUE #547: Get-DbaDatabaseLastUsed
Is this a *feature* OR *bug*:
Enhancement 

https://trello.com/c/s2fnvD0C

[Last_Access.zip](https://github.com/sqlcollaborative/dbatools/files/715295/Last_Access.zip)

**Additional comments**
@Jagoop 
"Use data from sys.dm_db_index_usage_stats to determine when the database was last used by users.
Max of last_user_seek, last_user_scan, last_user_lookup = last time a user read.
last_user_update = last time a user wrote/updated.

The method isn't foolproof, but combined with the database created timestamp and the instance restert (from tempdb creation date), it can assist in discovering databases that are unused."

@wsmelton 
"Reading "Watch-SqlDbLogin" it would seem this may be close or just another option to doing the same thing for databases. Watch command is named for logins but it can be the same premise around databases as well.

But this command could still prove useful. Could also add in something fancy like just doing a "Watch-DbaDatabase" instead and periodically poll the index usage stats....or create an Trace/XEvent to watch for login activity, then load that into table or display output."

### PR DESCRIPTION
**Pull request by SQLDBAWithABeard**
## Type of Change

Adds -LastUsed parameter to Get-DBADatabase returning the last index usage from sys.dm_db_index_usage_stats DMV
 - [X] Bug fix (non-breaking change, fixes #547
 - [X ] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
.\tests\manual.pester.ps1 -Show Summary
Tests completed in 196.88s
Tests Passed: 9897, Failed: 0, Skipped: 0, Pending: 0, Inconclusive: 0

 - [X ] Pester test is included
Invoke-Pester -Script .\tests\Get-DbaDatabase.Tests.ps1 -Tag Unittests -Show Summary
Tests Passed: 6, Failed: 0, Skipped: 0, Pending: 0, Inconclusive: 0
Added Unit test for NoSytemDbs and NoUserDbs
Added Unit Test for -LastUsed

### CODE REVIEW

**Comment by SQLDBAWithABeard** [comment]
Test failure is for Restore-DBFromFilteredArray 

Restore-DBFromFilteredArray - Microsoft.SqlServer.Management.Smo.FailedOperationException: Restore failed for Server 'APPVYR-WIN'.  ---> Microsoft.SqlServer.Management.Smo.SmoException: System.Data.SqlClient.SqlError: Transaction (Process ID 60) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction.

   at Microsoft.SqlServer.Management.Smo.ExecutionManager.ExecuteNonQueryWithMessage(StringCollection queries, ServerMessageEventHandler dbccMessageHandler, Boolean errorsAsMessages, Boolean retry)

   at Microsoft.SqlServer.Management.Smo.BackupRestoreBase.ExecuteSql(Server server, StringCollection queries)

   at Microsoft.SqlServer.Management.Smo.Restore.SqlRestore(Server srv)

   --- End of inner exception stack trace ---

   at Microsoft.SqlServer.Management.Smo.Restore.SqlRestore(Server srv)

   at CallSite.Target(Closure , CallSite , Object , Object )

See error log C:\Users\appveyor\Documents\dbatools-exceptions.txt for more details.

**Comment by potatoqualitee** [comment]
No problem, @SQLDBAWithABeard - that happens. I added some sleeps but can't yet figure out how to make it not deadlock.

**Comment by potatoqualitee** [comment]
Thanks so much for the PR! Will test again and merge today 💃

File: functions/Get-DbaDatabase.ps1 [hunk 1]
```diff
@@ -20,18 +23,17 @@
 		.PARAMETER ExcludeDatabase
 			The database(s) to exclude.
 
-		.PARAMETER NoUserDb
```
Comment (potatoqualitee) [inline functions/Get-DbaDatabase.ps1]: I updated this while I was all up in the file

File: functions/Get-DbaDatabase.ps1 [hunk 2]
```diff
@@ -159,58 +165,104 @@
...
+				   UNION ALL
+				   SELECT dbname, NULL, last_user_update FROM agg
+				) AS x (dbname, last_read, last_write)
+				GROUP BY
+				   dbname
+				ORDER BY 1;"
+			# put a function around this to enable Pester Testing and also to ease any future changes
+			function Invoke-QueryDBlastUsed {
+				try {
+					$server.Query($querylastused)
```
Comment (potatoqualitee) [inline functions/Get-DbaDatabase.ps1]: Fred added some awesome methods to $server and $database! So now we're using $server.Query() which outputs the tables thing like you were doing

File: functions/Get-DbaDatabase.ps1 [hunk 3]
```diff
@@ -220,18 +272,19 @@
...
-			$defaults =  'ComputerName', 'InstanceName', 'SqlInstance', 'Name', 'Status', 'IsAccessible', 'RecoveryModel',
-                         'LogReuseWaitStatus','Size as SizeMB', 'CompatibilityLevel as Compatibility', 'Collation', 'Owner',
-                         'LastBackupDate as LastFullBackup', 'LastDifferentialBackupDate as LastDiffBackup',
-                         'LastLogBackupDate as LastLogBackup'
-
+			
+			$defaults = 'ComputerName', 'InstanceName', 'SqlInstance', 'Name', 'Status', 'IsAccessible', 'RecoveryModel',
+			'LogReuseWaitStatus', 'Size as SizeMB', 'CompatibilityLevel as Compatibility', 'Collation', 'Owner',
+			'LastBackupDate as LastFullBackup', 'LastDifferentialBackupDate as LastDiffBackup',
+			'LastLogBackupDate as LastLogBackup', 'LastRead as LastIndexRead', 'LastWrite as LastIndexWrite'
```
Comment (potatoqualitee) [inline functions/Get-DbaDatabase.ps1]: Kept it as nospaces

File: functions/Get-DbaDatabase.ps1 [hunk 4]
```diff
@@ -240,16 +293,22 @@
...
 					}
-					Add-Member -Force -InputObject $db -MemberType NoteProperty BackupStatus -value $Notes
+					
+					$lastusedinfo = $dblastused | Where-Object { $_.dbname -eq $db.name }
 					
 					Add-Member -Force -InputObject $db -MemberType NoteProperty -Name ComputerName -value $server.NetName
 					Add-Member -Force -InputObject $db -MemberType NoteProperty -Name InstanceName -value $server.ServiceName
 					Add-Member -Force -InputObject $db -MemberType NoteProperty -Name SqlInstance -value $server.DomainInstanceName
+					Add-Member -Force -InputObject $db -MemberType NoteProperty -Name LastRead -value $lastusedinfo.last_read
+					Add-Member -Force -InputObject $db -MemberType NoteProperty -Name LastWrite -value $lastusedinfo.last_write
```
Comment (potatoqualitee) [inline functions/Get-DbaDatabase.ps1]: With 1.0, we'll be aiming to keep all columns returning all the time but only hiding them when needed. I like this feature enough, I included it by default all the time ❤️

File: functions/Get-DbaDatabase.ps1 [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
-	<#
+    <#
 		.SYNOPSIS
 			Gets SQL Database information for each database that is present in the target instance(s) of SQL Server.
 
 		.DESCRIPTION
 			The Get-DbaDatabase command gets SQL database information for each database that is present in the target instance(s) of
 			SQL Server. If the name of the database is provided, the command will return only the specific database information.
 
+			We also added a number of useful columns by default such as the Last used read and write Times for the Database using
```
Comment (PowerDBAKlaas) [inline functions/Get-DbaDatabase.ps1]: While it definitely is a useful addition to have extra properties and filtering options, this should not be a default behaviour.
Like a lot of functions, we get the SMO objects, fast and lean. Only when the user asks for more, we add functionality (and cost and time and risk).
There should be a switch to ask for this info, and without it, the query should not be executed.
@SQLDBAWithABeard can you make that like @niphlod did for the `-NoBackup` ?

File: functions/Get-DbaDatabase.ps1 (hunk 4 above)
Comment (PowerDBAKlaas) [inline functions/Get-DbaDatabase.ps1]: This is not 'unhiding' a column, this is additional functionality from multiple sources.

**Comment by potatoqualitee** [comment]
that was my fault, @PowerDBAKlaas. I really like this information - if it only adds, say 33 ms, can we keep?

File: functions/Get-DbaDatabase.ps1 (hunk 2 above)
Comment (SQLDBAWithABeard) [inline functions/Get-DbaDatabase.ps1]: sweet - need these on the wiki unless I am blind ;-) (which I am !!)

File: functions/Get-DbaDatabase.ps1 (hunk 5 above)
Comment (potatoqualitee) [inline functions/Get-DbaDatabase.ps1]: This is my fault!

Comment (SQLDBAWithABeard) [inline functions/Get-DbaDatabase.ps1]: That's what it was - Chrissy changed it - so whichever way is decided we can do :-)

Comment (potatoqualitee) [inline functions/Get-DbaDatabase.ps1]: Fineeeeeeeeeee just did a test and it added 300ms to my get 👎

File: functions/Get-DbaDatabase.ps1 (hunk 4 above)
Comment (potatoqualitee) [inline functions/Get-DbaDatabase.ps1]: i appreciate your pragmatism and have addressed your concerns 💯

**Comment by potatoqualitee** [comment]
@PowerDBAKlaas I have addressed your requests and the tests pass. plz approve

**Comment by PowerDBAKlaas** [comment]
nice.
Only, `-Lastused` makes me think I need to supply a date to it. As a switch, wouldn't `-IncludeLastUsed` or something be more clear?

**Comment by potatoqualitee** [comment]
that crossed my mind as well. i like the proposed name, will patch

**Pull request merged**